### PR TITLE
Make callbackList a linked-list instead of an array

### PR DESCRIPTION
**Pull request by acdlite**
See https://github.com/facebook/react/pull/8728#discussion_r95643611

### CODE REVIEW

File: src/renderers/shared/fiber/ReactFiberUpdateQueue.js [hunk 1]
```diff
@@ -28,6 +28,36 @@ const {
...
+  next: ListNode<T> | null,
+};
+
+type List<T> = {
+  first: ListNode<T> | null,
+  last: ListNode<T> | null,
+};
+
+function createList<A>() : List<A> {
+  return { first: null, last: null };
```
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberUpdateQueue.js]: Meh I guess I should just put these directly on the update queue type :D

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberUpdateQueue.js]: What's more important, that this List type is generic or that we avoid the extra allocation?

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberUpdateQueue.js]: Could also make the List type an intersection of the ListNode type, like the UpdateQueue type was in the first version.

Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberUpdateQueue.js]: Making it generic is explicitly a non-goal. If all we need is something super generic. Maybe we can just stick to arrays.

**Comment by acdlite** [comment]
@sebmarkbage Ok I put the list on the UpdateQueue.

**Comment by sebmarkbage** [comment]
I think my original feedback was too eager. I was thinking that we could reuse other parts of the linked list data structure for updates.

I also added more uses of dynamic arrays in #8607.

So maybe we can just leave it as is?

**Comment by acdlite** [comment]
Leave it as an array? Sure

**Pull request closed without merging**